pennylane_lightning/core/src/simulators/lightning_gpu/observables/ObservablesGPU.hpp

```diff
@@ -0,0 +1,212 @@
+    create(std::initializer_list<std::shared_ptr<Observable<StateVectorT>>> obs)
+        -> std::shared_ptr<TensorProdObs<StateVectorT>> {
+        return std::shared_ptr<TensorProdObs<StateVectorT>>{
+            new TensorProdObs(std::move(obs))};
+    }
+
+    static auto
+    create(std::vector<std::shared_ptr<Observable<StateVectorT>>> obs)
+        -> std::shared_ptr<TensorProdObs<StateVectorT>> {
+        return std::shared_ptr<TensorProdObs<StateVectorT>>{
+            new TensorProdObs(std::move(obs))};
+    }
+};
+
+/**
+ * @brief General Hamiltonian as a sum of observables.
+ *
+ */
+template <class StateVectorT>
+class Hamiltonian final : public HamiltonianBase<StateVectorT> {
+  private:
+    using BaseType = HamiltonianBase<StateVectorT>;
+
+  public:
+    using PrecisionT = typename StateVectorT::PrecisionT;
+    using ComplexT = typename StateVectorT::ComplexT;
+
+    /**
+     * @brief Create a Hamiltonian from coefficients and observables
+     *
+     * @param coeffs Arguments to construct coefficients
+     * @param obs Arguments to construct observables
+     */
+    template <typename T1, typename T2>
+    explicit Hamiltonian(T1 &&coeffs, T2 &&obs) : BaseType{coeffs, obs} {}
+
+    /**
+     * @brief Convenient wrapper for the constructor as the constructor does not
+     * convert the std::shared_ptr with a derived class correctly.
+     *
+     * This function is useful as std::make_shared does not handle
+     * brace-enclosed initializer list correctly.
+     *
+     * @param coeffs Argument to construct coefficients
+     * @param obs Argument to construct terms
+     */
+    static auto
+    create(std::initializer_list<PrecisionT> coeffs,
+           std::initializer_list<std::shared_ptr<Observable<StateVectorT>>> obs)
+        -> std::shared_ptr<Hamiltonian<StateVectorT>> {
+        return std::shared_ptr<Hamiltonian<StateVectorT>>(
+            new Hamiltonian<StateVectorT>{std::move(coeffs), std::move(obs)});
+    }
+
+    // to work with
+    void applyInPlace(StateVectorT &sv) const override {
+        using CFP_t = typename StateVectorT::CFP_t;
+        std::unique_ptr<DataBuffer<CFP_t>> buffer =
+            std::make_unique<DataBuffer<CFP_t>>(sv.getDataBuffer().getLength(),
+                                                sv.getDataBuffer().getDevTag());
+        buffer->zeroInit();
+
+        for (size_t term_idx = 0; term_idx < this->coeffs_.size(); term_idx++) {
+            StateVectorT tmp(sv);
+            this->obs_[term_idx]->applyInPlace(tmp);
+            scaleAndAddC_CUDA(
+                std::complex<PrecisionT>{this->coeffs_[term_idx], 0.0},
+                tmp.getData(), buffer->getData(), tmp.getLength(),
+                tmp.getDataBuffer().getDevTag().getDeviceID(),
+                tmp.getDataBuffer().getDevTag().getStreamID(),
+                tmp.getCublasCaller());
+        }
+        sv.updateData(std::move(buffer));
+    }
+};
+
+} // namespace Pennylane::LightningGPU::Observables
```

Comment:
Are we missing the Sparse addition here?

Reply:
As @vincentmr  mentioned above, sparse related computation are moved to measurement. There is no sparsehamiltonian for both lightning_qubit and LKokkos. So let's follow current design of lightning.

Reply:
And these both work with a sparse H in the adjoint pipeline, even with this design?

Reply:
yes, it works!

Reply:
Works fine!